### DISCUSSION TITLE
Split an image into lines of text

### DISCUSSION THREAD

**Discussion by jmspaggi** [ImageMagick/ImageMagick#2649] [Help] [answered]
Hi,

I have an image looking like this:
![image](https://user-images.githubusercontent.com/3017827/94697179-e06ea700-0305-11eb-9d48-32c212227573.png)
and I would like to split it into 4 images, one per line. I can do a small app to do that by looking at black lines, but I'm pretty sure there is some magick too that can already do it.

any idea how this can be achieved?

Thanks,

JMS

**Comment by fmw42** [comment, marked as answer]
Here is one way to do that using morphology and connected components. Use morphology to close up the rows of text. Then use connected components to get the bounding boxes and use those to crop the image.

![text](https://user-images.githubusercontent.com/13458218/94720382-43157200-0309-11eb-93bb-c69653fa9be0.png)


bboxes=`convert text.png -morphology close rectangle:30x10 -type bilevel \
-define connected-components:verbose=true \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-connected-components 4 x.png | grep "gray(255)" | awk '{print $2}'`
echo $bboxes
i=0
for box in $bboxes; do
convert text.png -crop $box +repage text_$i.png
i=$((i+1))
done


![text_0](https://user-images.githubusercontent.com/13458218/94721251-75739f00-030a-11eb-9e6c-be38f96b3b6b.png)


![text_1](https://user-images.githubusercontent.com/13458218/94721260-79072600-030a-11eb-8cf9-abdb39d2d7e0.png)

![text_2](https://user-images.githubusercontent.com/13458218/94721266-7b698000-030a-11eb-93ff-051b300f6dce.png)

![text_3](https://user-images.githubusercontent.com/13458218/94721279-7f959d80-030a-11eb-8897-1218bc3a27c5.png)

If you need them in proper order, they could be sorted by y before cropping.

**Reply by jmspaggi** [reply]
Ha, nice!!! 

I'm able to get the boxes and they seems to be correct:
`1005x60+18+113
566x55+21+25
508x49+21+187`
But the extraction of the picture seems to fail:
convert text.png -crop 1005x60+18+113 +repage text_0.png 
convert-im6.q16: geometry does not contain image `text.png' @ warning/transform.c/CropImage/667
Any idea why?
Also, is it possible to keep some of the black arround the boxes? The OCR doesn't like when the letters are touching the border :(

**Reply by fmw42** [reply]
You can add a border or extent with black in the crop command to add black all around or just on the sides.

Post your image so that I can see what is wrong and why you get that message. It seems to indicate that you have a region either outside the image or of zero size.

**Reply by jmspaggi** [reply]
The image is the one above :( That's why I'm confused.

The size seems to be correct:
identify -format "%[fx:w]x%[fx:h]" text.png

1075x287

But it still says that it does not contain... Weird.

Do you have the syntax for the border/extent?

I will keep trying with the crop to figure why it says that. 

Here is the image again. Hopefuly as a link.
![text](https://user-images.githubusercontent.com/3017827/94731289-8801ce80-0332-11eb-9538-f071066633e8.png)
(Ok. it's a different image ;) But result is the same. Boxes are the following: 1005x60+18+113 566x55+21+25 508x49+21+187)

Even this fails: convert text.png -crop 10x10+10+10 +repage text_0.png

**Reply by snibgo** [reply]
Your image has canvas offsets. I suggest you remove these with "+repage" immediately after reading it.

**Reply by jmspaggi** [reply]
Ha That works! How did you figured it had a canvas offset?

**Reply by snibgo** [reply]
Brute intelligence, or experience. You said:
> Even this fails: convert text.png -crop 10x10+10+10 +repage text_0.png [error message: geometry does not contain image]

I thought: how could that fail, with that error message? The image is clearly larger than 20x20. The answer is: It will fail if it has a canvas with offsets that don't cover the pixels your crop asks for. "magick identify" confirmed this:
```
x.png PNG 1075x287 1440x2960+163+250 8-bit Gray 2c 3876B 0.016u 0:00.064
```

**Reply by jmspaggi** [reply]
Interesting. I didn't even know this was possible!

**Comment by jmspaggi** [comment]
`identify -format "%[fx:w]x%[fx:h]" text.png 
1075x287`

**Comment by fmw42** [comment]
Your file has a virtual canvas offset (probably from a previous crop).  So you need to add +repage after reading the input. For example, this works for me:

```
bboxes=`convert text.png +repage -morphology close rectangle:30x10 -type bilevel \
-define connected-components:verbose=true \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-connected-components 4 x.png | grep "gray(255)" | awk '{print $2}'`
echo $bboxes
i=0
for box in $bboxes; do
convert text.png +repage -crop $box +repage text_$i.png
i=$((i+1))
done

```

Note the +repage in two places

**Reply by jmspaggi** [reply]
Awesome. Thanks a lot. That plus the border makes the OCR works like a charm!

**Reply by fmw42** [reply]
Sorry, for the border you can replace

`convert text.png +repage -crop $box +repage text_$i.png`

with

`convert text.png +repage -crop $box +repage -background black -border 10x0 text_$i.png`

Change the 10 to whatever value you want on the left and right sides each. The 0 says no border on the top and bottom. If you want border all around, use -border 10x10 or just -border 10

**Comment by fmw42** [comment]
I get 4:

```
bboxes=`convert text.png +repage -morphology close rectangle:30x10 -type bilevel \
-define connected-components:verbose=true \
-define connected-components:exclude-header=true \
-define connected-components:mean-color=true \
-connected-components 4 x.png | grep "gray(255)" | awk '{print $2}'`
echo $bboxes
i=0
for box in $bboxes; do
convert text.png +repage -crop $box +repage text_$i.png
i=$((i+1))
done

```
Check x.png and you will see if each row is fully connected.

Here are my outputs:

![text_0](https://user-images.githubusercontent.com/13458218/94837072-a1555a00-03c8-11eb-84ef-4dbf40ced6e8.png)

![text_1](https://user-images.githubusercontent.com/13458218/94837122-afa37600-03c8-11eb-98de-564b2b1b77c1.png)

![text_2](https://user-images.githubusercontent.com/13458218/94837083-a4e8e100-03c8-11eb-855b-0d464189c31e.png)

![text_3](https://user-images.githubusercontent.com/13458218/94837090-a74b3b00-03c8-11eb-995f-90e27eb6d745.png)

**Reply by jmspaggi** [reply]
Yes, works fine. My issue was different. Was my mistake.

Later, I figured that for some images, it splits the line into 2 pictures. increasing the rectangle size fixed that. 

Here is an example of a picture where I get 5 "lines" with a rectangle of 30x10 and 4 with using 40x10:
![name](https://user-images.githubusercontent.com/3017827/94837708-96a7be80-03e2-11eb-9c69-89d3d9921675.png)

I think with 40x10 it works very well. What is the "risk" of increasing this number? It will miss lines who are shorter than 40 pixels?

**Reply by fmw42** [reply]
Yes, it is up to you to set the morphology rectangle size to close the gaps appropriately.  Making it somewhat longer that needed should not hurt and might include more black on the sides. If you want the black on the sides, you can try changing "close" to "dilate", but then keep the height at 1 or small so as not to merge multiple lines
